feat(linter/plugins): comment-related APIs

lilnasy opened the pull request:
Part of #14564. Implement the remaining `SourceCode` APIs related to comments (`getCommentsBefore`, `getCommentsAfter`, `getCommentsInside`, `commentsExistBetween`).

lilnasy committed:
feat(linter/plugins): comment-related APIs

lilnasy commented on apps/oxlint/src-js/plugins/source_code.ts:
```diff
@@ -169,29 +169,66 @@ export const SOURCE_CODE = Object.freeze({
-  // oxlint-disable-next-line no-unused-vars
   getCommentsAfter(nodeOrToken: NodeOrToken): Comment[] {
-    throw new Error('`sourceCode.getCommentsAfter` not implemented yet'); // TODO
+    if (ast === null) initAst();
```
right, switching to draft

overlookmotel replied:
Don't take AI's reviews too seriously. Sometimes he catches things, but a lot of the time he talks bollocks.

In this case, the JSDoc comments were copied from ESLint. The word "directly" is odd (unnecessary), so by all means remove it if you like, but also feel free not to worry about it.

overlookmotel commented on apps/oxlint/src-js/plugins/source_code.ts:
```diff
@@ -169,29 +169,66 @@ export const SOURCE_CODE = Object.freeze({
+    if (comments.length === 0) return [];
+    const targetStart = nodeOrToken.range[0];
+    let i = comments.length - 1;
+    // linear backwards search for the index of the last comment before the provided node or token, TODO(perf): binary
```
Just a nit. Try to keep comments short where you can, and capitalize (I'm a capitalist!)

```suggestion
    // Linear backwards search for last comment before `nodeOrToken`.
    // TODO: Use binary search.
```

overlookmotel commented on apps/oxlint/src-js/plugins/source_code.ts:
```diff
@@ -169,29 +169,66 @@ export const SOURCE_CODE = Object.freeze({
+    const targetEnd = nodeOrToken.range[1];
+    let i = 0;
+    // linear forwards search for the index of the first comment after the provided node or token, TODO(perf): binary
+    while (i < comments.length && comments[i].range[0] < targetEnd) {
```
```suggestion
    const len = comments.length;
    while (i < len && comments[i].range[0] < targetEnd) {
```

...to avoid `comments.length` property lookup on each turn of the loop.

overlookmotel reviewed (commented):
Apologies that this review is premature - I know you're still working on it. But I was interested and had a look, so I thought I'd give my thoughts.

Background: In linter JS plugins, the JS-side code is going to be the slowest code in the whole path, so I'm keen to get it as tight and performant as it possibly can be. In most projects, readability and suchlike is also important, and it's normal to be a little loose, but here I think we should prioritise perf over all else.

(e.g. you'll see in the rest of the JS plugins code we never use `if (!foo)`, we'll use the cheaper `if (foo === false)` or `if (foo !== null)` instead, depending on the type of `foo`)

That's the reason for my rather fanatical nit-picking!

Note: I think it's fine to leave binary search to a later PR, but I still feel it's helpful to tighten the algorithms as much as possible at this earlier stage.

lilnasy committed:
ensure adjacency in  and

lilnasy committed:
style

lilnasy committed:
refactor: use whitespace regex as a constant

lilnasy committed:
refactor : no branching, collect slice start and end

lilnasy committed:
refactor : no branching

lilnasy committed:
refactor : remove early return

overlookmotel commented:
OK, I think I've recovered the commits I over-wrote, and force pushed them back onto the PR branch. But can you please check I didn't lose anything in the process? Really sorry I've made a mess of this.

overlookmotel committed:
Disable built-in rules in test

To make the snapshot easier to read...

overlookmotel committed:
Use `range` for nodes

overlookmotel committed:
Avoid array destructuring

Destructuring treats array as an iterator. Direct property access may be faster.

overlookmotel committed:
Move `commentsExistBetween` up to with other comments methods

overlookmotel committed:
Remove temp var in `commentsExistBetween`

If there's no comments after `betweenRangeStart`, `betweenRangeEnd` is never accessed. So move it to where it's used.

overlookmotel committed:
Avoid property lookups on each turn of loop

overlookmotel committed:
Re-order code

overlookmotel committed:
fix: Init `ast` in `commentsExistBetween`

overlookmotel commented:
I'm still going through it. Have pushed a few commits - all just nits, except the last one.

lilnasy committed:
rename `indexStart` to `sliceStart` with comment

overlookmotel committed:
Don't use `definePlugin` in test

overlookmotel committed:
Test output comments before/inside/after for all var and function decls

overlookmotel committed:
Add comment inside var decl in test fixture

overlookmotel committed:
More tests for `commentsExistBetween`

overlookmotel committed:
Update values of var decls in test fixture

overlookmotel committed:
Test where comments before and after is empty array

overlookmotel committed:
Rename test fixture file

overlookmotel committed:
Add test fixture containing no comments

overlookmotel committed:
Test behavior of `commentsExistBetween` when out of order

overlookmotel committed:
Test statement with no comments after

overlookmotel committed:
Rename var

overlookmotel committed:
Remove temp vars

Use comments instead for clarity.

overlookmotel reviewed (approved):
Nice one! I had completely misunderstood what these methods are meant to do, because I hadn't looked at ESLint's implementation. Thank you for figuring out the correct logic.

I've pushed a bunch of commits, but they're all just style nits, perf micro-micro-optimizations, and increasing the test coverage. The logic is unaltered, and all seems correct.

I've not checked the output against ESLint, but I have been through the snapshot and checked it's correct according to my new understanding of what these methods are meant to do.

The only one of my commits which actually changed behavior is this one: 1aadc6d78c8e1e94f692f175b0a68427977f1614

Going to merge now, but please come back if you disagree with any of my changes.

overlookmotel commented on apps/oxlint/src-js/plugins/source_code.ts:
```diff
@@ -170,29 +171,115 @@ export const SOURCE_CODE = Object.freeze({
+      const gap = sourceText.slice(targetEnd, commentStart);
+      const whitespaceOnlyGap = whitespacePattern.test(gap);
+      if (whitespaceOnlyGap) {
+        commentsAfter.push(comment);
```
Thanks for checking. Interesting!

I think it's fine to leave it as is for now. The larger perf improvement will be from switching from linear to binary search, and that may alter how this needs to work anyway.

overlookmotel commented on apps/oxlint/src-js/plugins/source_code.ts:
```diff
@@ -169,29 +169,66 @@ export const SOURCE_CODE = Object.freeze({
+    const targetStart = nodeOrToken.range[0];
+    let i = comments.length - 1;
+    // linear backwards search for the index of the last comment before the provided node or token, TODO(perf): binary
+    while (i >= 0 && comments[i].range[1] > targetStart) {
```
Thanks for making that change. I've reverted to using `range` for `nodeOrToken`, as that's user-supplied and the shape of the object they pass in is unknown. I think ESLint tends to use `range`.
